### PR TITLE
Cherry-pick fixes from master into 2.0

### PR DESCRIPTION
**Pull request by 23Skidoo**
@hvr, please ping me if I have forgotten something.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Not sure if @dcoutts's concurrent store stuff should go in 2.0 at this stage, what do others think?

**Comment by ezyang** [comment]
It should, because the cabal-install code is low risk, we can always cut a new release.

**Comment by 23Skidoo** [comment]
Added Duncan's file locking stuff and Edward's AppVeyor fixes.

**Comment by ezyang** [comment]
@23Skidoo Why don't we bite the bullet and cherry pick the Travis timeout fixes to 2.0 as well? That will help 2.0 CIs a bit.

**Comment by 23Skidoo** [comment]
I'm all for it.

File: travis-script.sh [hunk 1]
```diff
@@ -149,6 +149,7 @@ fi
 
 timed cabal new-build $jobs $CABAL_INSTALL_FLAGS \
                       cabal-install:cabal \
+                      cabal-install:integration-tests \
```
Comment (23Skidoo) [inline travis-script.sh]: They were actually removed in 2.0 as well (9368717f919326b0dcd3e03bbe43d4aa99624ef8).

**Comment by 23Skidoo** [comment]
> ```
> Shutting down GHCi sessions (please be patient)...
> Unexpected failure on GHCi exit: fd:6: hClose: resource vanished (Broken pipe)
> cabal-tests.exe: fd:6: hClose: resource vanished (Broken pipe)
> ```

Anyone seen this before?

**Comment by ezyang** [comment]
Seems to be OK now?

**Comment by 23Skidoo** [comment]
Yep, I restarted the build.

**Comment by ezyang** [comment]
Oh, did you fix the integration-tests call? We need to stop running it from the test script.

**Comment by 23Skidoo** [comment]
Yes, I did.

**Comment by ezyang** [comment]
integration-tests call from travis-tests.sh needs to be removed.

**Comment by 23Skidoo** [comment]
It's green, merging.

**Pull request merged**